### PR TITLE
Make it possible to trim custom attribute artifacts

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
Pretty much any app uses custom attribute metadata due to `FlagsAttribute`. But not every app needs to activate custom attributes. We currently root unreachable code when custom attributes are not actually reflected on (such as constructors, property setters, MethodTables, etc.). The code reachable from there has potentially a big closure. This adds a check to make sure we only do this if there's a chance the attributes are looked at.

### CODE REVIEW

**Comment by jkotas** [comment]
Are there examples of where this kicks in and how much does it save?

**Comment by MichalStrehovsky** [comment]
> Are there examples of where this kicks in and how much does it save?

This is a stepping stone. I'm trying to get rid of boxed enums (e.g. `MethodTable` for `AttributeTargets` enum).

`MethodTable` for `Enum` has a ton of dependencies due to all the generic specialization that was done to make stringification fast.

https://github.com/MichalStrehovsky/rt-sz/issues/10
https://github.com/MichalStrehovsky/rt-sz/issues/9

Has the stats for the two pull requests I opened.

If I combine the two pull requests with the below patch (haven't really decided how I can make that one look more acceptable), we can get hello-minimal to 801,280 bytes (so getting rid of Enums with the two pull requests and this patch is almost 10% saving, but we need all of this to kick in to reap the benefit).

Maybe this can wait for whether the below diff can be shaped into something acceptable.

```diff
diff --git a/src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/MethodTable.Runtime.cs b/src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/MethodTable.Runtime.cs
index 899ac448d5f..c0c196713aa 100644
--- a/src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/MethodTable.Runtime.cs
+++ b/src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/MethodTable.Runtime.cs
@@ -37,7 +37,7 @@ internal bool IsEnum
 
                 // Generic type definitions that return true for IsPrimitive are type definitions of generic enums.
                 // Otherwise check the base type.
-                return IsPrimitive && (IsGenericTypeDefinition || NonArrayBaseType == MethodTable.Of<Enum>());
+                return IsPrimitive && (IsGenericTypeDefinition || NonArrayBaseType != MethodTable.Of<ValueType>());
             }
         }
 
diff --git a/src/coreclr/nativeaot/System.Private.CoreLib/src/System/GC.NativeAot.cs b/src/coreclr/nativeaot/System.Private.CoreLib/src/System/GC.NativeAot.cs
index 0ee16609157..8263fc7cab8 100644
--- a/src/coreclr/nativeaot/System.Private.CoreLib/src/System/GC.NativeAot.cs
+++ b/src/coreclr/nativeaot/System.Private.CoreLib/src/System/GC.NativeAot.cs
@@ -777,8 +777,8 @@ public static GCMemoryInfo GetGCMemoryInfo(GCKind kind)
                 throw new ArgumentOutOfRangeException(nameof(kind),
                                       SR.Format(
                                           SR.ArgumentOutOfRange_Bounds_Lower_Upper,
-                                          GCKind.Any,
-                                          GCKind.Background));
+                                          nameof(GCKind.Any),
+                                          nameof(GCKind.Background)));
             }
 
             var data = new GCMemoryInfoData();
diff --git a/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/Thread.NativeAot.Windows.cs b/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/Thread.NativeAot.Windows.cs
index d6ea54412e1..8d8d86215b3 100644
--- a/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/Thread.NativeAot.Windows.cs
+++ b/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/Thread.NativeAot.Windows.cs
@@ -289,7 +289,12 @@ private bool SetApartmentStateUnchecked(ApartmentState state, bool throwOnError)
             {
                 if (throwOnError)
                 {
-                    string msg = SR.Format(SR.Thread_ApartmentState_ChangeFailed, retState);
+                    string msg = SR.Format(SR.Thread_ApartmentState_ChangeFailed, retState switch
+                    {
+                        ApartmentState.STA => "STA",
+                        ApartmentState.MTA => "MTA",
+                        _ => "Unknown",
+                    });
                     throw new InvalidOperationException(msg);
                 }
 
diff --git a/src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs b/src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs
index 189ce06de70..b7f814ccf72 100644
--- a/src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs
+++ b/src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs
@@ -628,7 +628,7 @@ private static CultureData CreateCultureWithInvariantData()
             invariant._iFirstWeekOfYear = 0;                      // first week of year
 
             // all available calendar type(s).  The first one is the default calendar
-            invariant._waCalendars = new CalendarId[] { CalendarId.GREGORIAN };
+            invariant._waCalendars = (CalendarId[])(object)new ushort[] { (ushort)CalendarId.GREGORIAN };
 
             if (!GlobalizationMode.Invariant)
             {
diff --git a/src/libraries/System.Private.CoreLib/src/System/Globalization/GregorianCalendar.cs b/src/libraries/System.Private.CoreLib/src/System/Globalization/GregorianCalendar.cs
index edb65a80557..87092870126 100644
--- a/src/libraries/System.Private.CoreLib/src/System/Globalization/GregorianCalendar.cs
+++ b/src/libraries/System.Private.CoreLib/src/System/Globalization/GregorianCalendar.cs
@@ -50,8 +50,8 @@ public GregorianCalendar(GregorianCalendarTypes type)
             {
                 throw new ArgumentOutOfRangeException(
                     nameof(type),
-                    type,
-                    SR.Format(SR.ArgumentOutOfRange_Range, GregorianCalendarTypes.Localized, GregorianCalendarTypes.TransliteratedFrench));
+                    (int)type,
+                    SR.Format(SR.ArgumentOutOfRange_Range, nameof(GregorianCalendarTypes.Localized), nameof(GregorianCalendarTypes.TransliteratedFrench)));
             }
 
             _type = type;

```

**Comment by jkotas** [comment]
How does GetGCMemoryInfo get pulled into console hello world?

**Comment by MichalStrehovsky** [comment]
> How does GetGCMemoryInfo get pulled into console hello world?

This one looks easily fixable by just skipping the validation of the enum range since we know we used a valid enum.

It's from string interpolation:

image

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/CustomAttributeMetadataNode.cs [hunk 1]
```diff
@@ -3,7 +3,10 @@
 
 using System.Collections.Generic;
 
+using Internal.TypeSystem;
+
 using ILCompiler.DependencyAnalysisFramework;
+using System;
```
Comment (jkotas) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/CustomAttributeMetadataNode.cs:9]: Nit: Sort usings

**Comment by jkotas** [comment]
> Maybe this can wait for whether the below diff can be shaped into something acceptable.

I have been looking at this one on and off. My worry with this diff is that requires writing unnatural code in random places. It means that we would likely see regressions regularly and that the benefits are unlikely to apply to anything beyond the trivial `Console.WriteLine("Hello world")`. I am not sure we want to go ahead with a change like that. Though, I agree it would be nice to make the enum formatting more pay-for-play.

**Pull request closed without merging**